# &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; How technologies changed movies industry

The only constraints:

- you must use each and every method from movielens_analysis.py except class Tests
- every cell in your notebook should contain magic command %timeit
- all other imports are prohibited as well as using built-in functions. If you need them, put them in your module in advance

In [1]:
%timeit from movielens_analysis import Tags, Links, Ratings, Movies, Graph
from movielens_analysis import Tags, Links, Ratings, Movies, Graph

The slowest run took 8.01 times longer than the fastest. This could mean that an intermediate result is being cached.
1.62 μs ± 1.75 μs per loop (mean ± std. dev. of 7 runs, 1 loop each)


## &emsp;&emsp;&emsp;&emsp; Movie Trends & Genre Analysis

### To start with, analyze cinematography trends by years. Take a look at periods covered by dataset (1902 - 2018) and find the years in which the most movies were released.

In [2]:
%timeit Movies().dist_by_release()
mv = Movies().dist_by_release()
mv

20.2 ms ± 2.08 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


{'2002': 311,
 '2006': 295,
 '2001': 294,
 '2007': 284,
 '2000': 283,
 '2009': 282,
 '2003': 279,
 '2004': 279,
 '2014': 278,
 '1996': 276,
 '2015': 274,
 '2005': 273,
 '2008': 269,
 '1999': 263,
 '1997': 260,
 '1995': 259,
 '1998': 258,
 '2011': 254,
 '2010': 247,
 '2013': 239,
 '1994': 237,
 '2012': 233,
 '2016': 218,
 '1993': 198,
 '1992': 167,
 '1988': 165,
 '1987': 153,
 '1990': 147,
 '1991': 147,
 '2017': 147,
 '1989': 142,
 '1986': 139,
 '1985': 126,
 '1984': 101,
 '1981': 92,
 '1980': 89,
 '1982': 87,
 '1983': 83,
 '1979': 69,
 '1977': 63,
 '1973': 59,
 '1978': 59,
 '1965': 47,
 '1971': 47,
 '1974': 45,
 '1976': 44,
 '1964': 43,
 '1967': 42,
 '1968': 42,
 '1975': 42,
 '1966': 42,
 '2018': 41,
 '1962': 40,
 '1972': 39,
 '1963': 39,
 '1959': 37,
 '1960': 37,
 '1955': 36,
 '1969': 35,
 '1961': 34,
 '1970': 33,
 '1957': 33,
 '1958': 31,
 '1953': 30,
 '1956': 30,
 '1940': 25,
 '1949': 25,
 '1954': 23,
 '1942': 23,
 '1939': 23,
 '1946': 23,
 '1951': 22,
 '1950': 21,
 '1947': 20,
 '19

### Sustainable growth ending with plateau (in 2002 - 2015) is observed. Acceleration of technologies development is the probable reason of that.

### Let's examine the distribution by genres. Histogram shows that **drama** and **comedy** strongly lead over the others genres. 

In [3]:
%timeit Graph().text_histogram(Movies().dist_by_genres(),400)
histogram = Graph().text_histogram(Movies().dist_by_genres(),400)
histogram


17.3 ms ± 379 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


['Drama                ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 4361',
 'Comedy               ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 3756',
 'Thriller             ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 1894',
 'Action               ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 1828',
 'Romance              ■■■■■■■■■■■■■■■■■■■■■■■■■■■■■ 1596',
 'Adventure            ■■■■■■■■■■■■■■■■■■■■■■■ 1263',
 'Crime                ■■■■■■■■■■■■■■■■■■■■■ 1199',
 'Sci-Fi               ■■■■■■■■■■■■■■■■■ 980',
 'Horror               ■■■■■■■■■■■■■■■■■ 978',
 'Fantasy              ■■■■■■■■■■■■■■ 779',
 'Children             ■■■■■■■■■■■■ 664',
 'Animation            ■■■■■■■■■■■ 611',
 'Mystery              ■■■■■■■■■■ 573',
 'Documentary          ■■■■■■■■ 440',
 'War                  ■■■■■■■ 382',
 'Musical              ■■■■■■ 334',
 'Western              ■■■ 167',
 'IMAX                 ■■ 158',
 'Film-Noir            ■ 87',
 '(no genres listed)    34']


### It could be explained by the fact that comedy and drama remain the most popular genres due to their long-established history and broader audience appeal. In contrast, hybrid and niche genres (such as noir) demonstrate lower popularity, likely because of their specialized narratives and limited mainstream recognition.

### Then, investigate how many genres is the most typical for movies from dataset.

In [4]:
%timeit Movies().most_genres(1000)
most_gn = Movies().most_genres(1000)
most_gn

2.22 s ± 146 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


{'Rubber (2010)': 10,
 'Patlabor: The Movie (Kidô keisatsu patorebâ: The Movie) (1989)': 8,
 'Mulan (1998)': 7,
 'Who Framed Roger Rabbit? (1988)': 7,
 'Osmosis Jones (2001)': 7,
 'Interstate 60 (2002)': 7,
 'Robots (2005)': 7,
 'Pulse (2006)': 7,
 'Aqua Teen Hunger Force Colon Movie Film for Theaters (2007)': 7,
 'Enchanted (2007)': 7,
 'Aelita: The Queen of Mars (Aelita) (1924)': 7,
 'Inception (2010)': 7,
 'Tangled (2010)': 7,
 'Mars Needs Moms (2011)': 7,
 'Strange Days (1995)': 6,
 'The Lion King (1994)': 6,
 'The Getaway (1994)': 6,
 'Super Mario Bros. (1993)': 6,
 'Beauty and the Beast (1991)': 6,
 'All Dogs Go to Heaven 2 (1996)': 6,
 'Space Jam (1996)': 6,
 'Aladdin and the King of Thieves (1996)': 6,
 'Lost Highway (1997)': 6,
 'The Return of Jafar (1994)': 6,
 'Young Sherlock Holmes (1985)': 6,
 'The Mummy (1999)': 6,
 'Shrek (2001)': 6,
 'The Stunt Man (1980)': 6,
 'Laputa: Castle in the Sky (Tenkû no shiro Rapyuta) (1986)': 6,
 'Ichi the Killer (Koroshiya 1) (2001)': 6,
 '

### Majority of movies are characterised by typical range of 4-6 genres. Some outliers among the results are observed. It could be explained by the experimental nature of the "Rubber" movie (which is a surreal horror-comedy about a sentient tire) and hybrid nature of the mecha anime "Patlabor: The Movie". 

## &emsp;&emsp;&emsp;&emsp; Audience Ratings & Engagement

### Next, using Ratings.Movies, examine how ratings are distributed by year (dist_by_year)

In [5]:
%timeit Graph().text_histogram(Ratings().Movies().dist_by_year(),100,0)
rt_by_year = Ratings().Movies().dist_by_year()
histogram = Graph().text_histogram(rt_by_year,100,0)
histogram

203 ms ± 15.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


['2018   ■■■■■■■■■■■■ 6418',
 '2017   ■■■■■■■■■■■■■■■■ 8199',
 '2016   ■■■■■■■■■■■■■ 6702',
 '2015   ■■■■■■■■■■■■■ 6616',
 '2014   ■■ 1439',
 '2013   ■■■ 1664',
 '2012   ■■■■■■■■■ 4657',
 '2011   ■■■ 1690',
 '2010   ■■■■ 2300',
 '2009   ■■■■■■■■ 4158',
 '2008   ■■■■■■■■ 4351',
 '2007   ■■■■■■■■■■■■■■ 7114',
 '2006   ■■■■■■■■ 4059',
 '2005   ■■■■■■■■■■■ 5813',
 '2004   ■■■■■■ 3279',
 '2003   ■■■■■■■ 4014',
 '2002   ■■■■■■ 3478',
 '2001   ■■■■■■■ 3922',
 '2000   ■■■■■■■■■■■■■■■■■■■■ 10061',
 '1999   ■■■■ 2439',
 '1998   ■ 507',
 '1997   ■■■ 1916',
 '1996   ■■■■■■■■■■■■ 6040']

### The histogram shows three major peaks of users' activity amongst the years analyzed: 2000, 2007 and 2017. The first peak could be explained by the popularity of the MovieLens platform in the early 2000s. The second peak could be associated with growing interest in online cinemas and referral services in the mid-2000s, whereas the third one is probably linked with increase in development of streaming services (like Netflix).

### Despite all these explainations, we could conclude that user engagement on MovieLens was significantly influenced by multiple interdependent factors: technological infrastructure evolution, cultural and cinematic Trends, etc.

### Next step is to examine how ratings are distributed by marks (dist_by_rating)

In [6]:
%timeit Graph().text_histogram(Ratings().Movies().dist_by_rating(), 100)
rt_by_rating = Ratings().Movies().dist_by_rating()
histogram = Graph().text_histogram(rt_by_rating, 100)
histogram

59.1 ms ± 6.68 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


['4.0   ■■■■■■■■■■■■■■■■■■■■ 26818',
 '3.0   ■■■■■■■■■■■■■■ 20047',
 '5.0   ■■■■■■■■■ 13211',
 '3.5   ■■■■■■■■■ 13136',
 '4.5   ■■■■■■ 8551',
 '2.0   ■■■■■ 7551',
 '2.5   ■■■■ 5550',
 '1.0   ■■ 2811',
 '1.5   ■ 1791',
 '0.5   ■ 1370']

### People tend to give more average ratings(3.0 - 4.0) while more radical ones (especially negative - 0.5 - 1.5) are less common.


### Analyze top-movies by number of ratings (top_by_num_of_ratings) 

In [7]:
%timeit Ratings().Movies().top_by_num_of_ratings(1000)
top_10rt_by_year = Ratings().Movies().top_by_num_of_ratings(1000)
top_10rt_by_year

94.5 ms ± 8.37 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


{'Forrest Gump (1994)': 329,
 'The Shawshank Redemption (1994)': 317,
 'Pulp Fiction (1994)': 307,
 'The Silence of the Lambs (1991)': 279,
 'The Matrix (1999)': 278,
 'Star Wars: Episode IV - A New Hope (1977)': 251,
 'Jurassic Park (1993)': 238,
 'Braveheart (1995)': 237,
 'Terminator 2: Judgment Day (1991)': 224,
 "Schindler's List (1993)": 220,
 'Fight Club (1999)': 218,
 'Toy Story (1995)': 215,
 'Star Wars: Episode V - The Empire Strikes Back (1980)': 211,
 'The Usual Suspects (1995)': 204,
 'American Beauty (1999)': 204,
 'Seven (a.k.a. Se7en) (1995)': 203,
 'Independence Day (a.k.a. ID4) (1996)': 202,
 'Apollo 13 (1995)': 201,
 'Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) (1981)': 200,
 'The Lord of the Rings: The Fellowship of the Ring (2001)': 198,
 'Star Wars: Episode VI - Return of the Jedi (1983)': 196,
 'The Godfather (1972)': 192,
 'The Fugitive (1993)': 190,
 'Batman (1989)': 189,
 'Saving Private Ryan (1998)': 188,
 'The Lord of the Rings: T


### Cult and popular movies have the highest number of ratings. Critically acclaimed movies (e.g., "Shawshank Redemption") are also on top of the list.

### Let’s analyze their average and median rating (top_by_ratings with two metrics)

In [8]:
%timeit Ratings().Movies().top_by_ratings(metric='average')
top_10rt_by_average = Ratings().Movies().top_by_ratings(metric='average')
top_10rt_by_average

209 ms ± 8.42 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


{'Lamerica (1994)': 5.0,
 'Heidi Fleiss: Hollywood Madam (1995)': 5.0,
 'Awfully Big Adventure, An (1995)': 5.0,
 'Live Nude Girls (1995)': 5.0,
 'In the Realm of the Senses (Ai no corrida) (1976)': 5.0,
 'What Happened Was... (1994)': 5.0,
 'Thin Line Between Love and Hate, A (1996)': 5.0,
 'Denise Calls Up (1995)': 5.0,
 'Supercop 2 (Project S) (Chao ji ji hua) (1993)': 5.0,
 'Entertaining Angels: The Dorothy Day Story (1996)': 5.0,
 'Lesson Faust (1994)': 5.0,
 'Hype! (1996)': 5.0,
 'Vampire in Venice (Nosferatu a Venezia) (Nosferatu in Venice) (1986)': 5.0,
 'The Assignment (1997)': 5.0,
 'Four Days in September (O Que É Isso, Companheiro?) (1997)': 5.0,
 'Mephisto (1981)': 5.0,
 'Knock Off (1998)': 5.0,
 'The Ballad of Narayama (Narayama bushiko) (1983)': 5.0,
 'On the Ropes (1999)': 5.0,
 'Man and a Woman, A (Un homme et une femme) (1966)': 5.0,
 'Red Sorghum (Hong gao liang) (1987)': 5.0,
 'The Sandpiper (1965)': 5.0,
 'Babes in Toyland (1934)': 5.0,
 'My Man Godfrey (1957)': 5.

In [9]:
%timeit Ratings().Movies().top_by_ratings(metric='median')
top_10rt_by_median = Ratings().Movies().top_by_ratings(metric='median')
top_10rt_by_median

170 ms ± 10.3 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


{'Lamerica (1994)': 5.0,
 'Angels and Insects (1995)': 5.0,
 'Heidi Fleiss: Hollywood Madam (1995)': 5.0,
 'Awfully Big Adventure, An (1995)': 5.0,
 'Live Nude Girls (1995)': 5.0,
 'In the Realm of the Senses (Ai no corrida) (1976)': 5.0,
 'What Happened Was... (1994)': 5.0,
 'Thin Line Between Love and Hate, A (1996)': 5.0,
 'Denise Calls Up (1995)': 5.0,
 'Supercop 2 (Project S) (Chao ji ji hua) (1993)': 5.0,
 'Secrets & Lies (1996)': 5.0,
 'Children of the Corn IV: The Gathering (1996)': 5.0,
 'Entertaining Angels: The Dorothy Day Story (1996)': 5.0,
 'Lesson Faust (1994)': 5.0,
 'Hype! (1996)': 5.0,
 'Vampire in Venice (Nosferatu a Venezia) (Nosferatu in Venice) (1986)': 5.0,
 'The Assignment (1997)': 5.0,
 'Four Days in September (O Que É Isso, Companheiro?) (1997)': 5.0,
 'Mephisto (1981)': 5.0,
 'Knock Off (1998)': 5.0,
 "Swept Away (Travolti da un insolito destino nell'azzurro mare d'Agosto) (1975)": 5.0,
 'The Long Goodbye (1973)': 5.0,
 'The Ballad of Narayama (Narayama bushi

### The high score on both metrics (both average and median) does not reflect the film’s popularity. On top good movies with low rating are presented.

### To find controversial films, which drew attention a large number of various users, we could assess the variance of ratings (top_controversial):

In [10]:
%timeit Ratings().Movies().top_controversial(10)
contr_mv = Ratings().Movies().top_controversial(10)
contr_mv

166 ms ± 11.8 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


{'Fanny and Alexander (Fanny och Alexander) (1982)': 10.12,
 "Ivan's Childhood (a.k.a. My Name is Ivan) (Ivanovo detstvo) (1962)": 10.12,
 'Lassie (1994)': 8.0,
 'Zed & Two Noughts, A (1985)': 8.0,
 'Kwaidan (Kaidan) (1964)': 8.0,
 'Emma (2009)': 8.0,
 'Troll 2 (1990)': 6.75,
 'The Clonus Horror (1979)': 6.12,
 "To Grandmother's House We Go (1992)": 6.12,
 'The Purge: Election Year (2016)': 6.12}

In [11]:
%timeit Links().get_imdb(Links().csv_init_for_dct(Ratings().Movies().top_controversial(10, isreturnedid=1)), ['movieId', 'Title', 'Director', 'Budget', 'Cumulative Worldwide Gross', 'Runtime'])

contr_mv_id = Ratings().Movies().top_controversial(10, isreturnedid=1)
id_lst = Links().csv_init_for_dct(contr_mv_id)
info = Links().get_imdb(id_lst, ['movieId', 'Title', 'Director', 'Budget', 'Cumulative Worldwide Gross', 'Runtime'])
info

527 ms ± 21.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


[['84847', 'Emma', 'N/A', 'None', 'None', '240 min'],
 ['7564', 'Kwaidan', 'Masaki Kobayashi', '0', '0', '183 min'],
 ['70946', 'Troll 2', 'Claudio Fragasso', '65000', '0', '95 min'],
 ['484', 'Lassie', 'Daniel Petrie', '0', '9979683', '94 min'],
 ['32892',
  "Ivan's Childhood",
  'Andrei Tarkovsky, Eduard Abalov',
  '0',
  '0',
  '95 min'],
 ['3223', 'A Zed & Two Noughts', 'Peter Greenaway', '0', '0', '115 min'],
 ['26409', 'The Clonus Horror', 'Robert S. Fiveson', '257000', '0', '90 min'],
 ['2068',
  'Fanny and Alexander',
  'Ingmar Bergman',
  '6000000',
  '6800000',
  '188 min'],
 ['160565',
  'The Purge: Election Year',
  'James DeMonaco',
  '10000000',
  '118587880',
  '108 min'],
 ['121097',
  "To Grandmother's House We Go",
  'Jeff Franklin',
  '0',
  '0',
  '96 min']]

### From the information about the most controversial movies we could notice notable directors and art films. 
### Andrei Tarkovsky (Ivan's Childhood) and Ingmar Bergman (Fanny and Alexander) are titans of European art cinema, known for their philosophical depth and visual poetry.

### Peter Greenaway (A Zed & Two Noughts) and Masaki Kobayashi (Kwaidan) are also celebrated auteurs, with Kwaidan being a visually stunning Japanese anthology of ghost stories. 

### These films are often studied in film theory/criticism circles for their artistic innovation.

## &emsp;&emsp;&emsp;&emsp;  User Behavior Insights

### Next, using the class Users, we analyze user activity: how many ratings did they put? (similar to dist_by_year?)

In [12]:
%timeit Ratings().Users().dist_by_rating()
rt_by_users = Ratings().Users().dist_by_rating()
rt_by_users

75.6 ms ± 7.96 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


{'414': 2698,
 '599': 2478,
 '474': 2108,
 '448': 1864,
 '274': 1346,
 '610': 1302,
 '68': 1260,
 '380': 1218,
 '606': 1115,
 '288': 1055,
 '249': 1046,
 '387': 1027,
 '182': 977,
 '307': 975,
 '603': 943,
 '298': 939,
 '177': 904,
 '318': 879,
 '232': 862,
 '480': 836,
 '608': 831,
 '600': 763,
 '483': 728,
 '590': 728,
 '105': 722,
 '19': 703,
 '305': 677,
 '489': 648,
 '111': 646,
 '438': 635,
 '217': 613,
 '140': 608,
 '477': 600,
 '555': 578,
 '91': 575,
 '28': 570,
 '219': 528,
 '534': 520,
 '89': 518,
 '64': 517,
 '226': 507,
 '561': 505,
 '18': 502,
 '525': 500,
 '57': 476,
 '381': 474,
 '368': 469,
 '509': 467,
 '469': 465,
 '560': 458,
 '462': 455,
 '292': 446,
 '21': 443,
 '597': 443,
 '42': 440,
 '160': 437,
 '294': 437,
 '580': 436,
 '596': 411,
 '202': 403,
 '275': 403,
 '517': 400,
 '45': 399,
 '156': 398,
 '514': 397,
 '391': 386,
 '567': 385,
 '357': 383,
 '103': 377,
 '339': 371,
 '62': 366,
 '199': 363,
 '125': 360,
 '51': 359,
 '132': 347,
 '66': 345,
 '313': 340,
 

### Top-1 User is the one with id 414, who rated movies 2698 times.

### Let us examine user engagement patterns by quantifying the central tendency of their rating distributions, specifically through measures of arithmetic mean (average) and median rating values. 

In [13]:
%timeit Ratings().Users().top_by_ratings(metric='average')
top_users_by_average = Ratings().Users().top_by_ratings(metric='average')
top_users_by_average

139 ms ± 8.82 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


{'53': 5.0,
 '251': 4.87,
 '515': 4.85,
 '25': 4.81,
 '30': 4.74,
 '523': 4.69,
 '348': 4.67,
 '171': 4.63,
 '452': 4.56,
 '43': 4.55,
 '122': 4.55,
 '371': 4.55,
 '441': 4.52,
 '400': 4.51,
 '52': 4.48,
 '538': 4.47,
 '168': 4.46,
 '417': 4.46,
 '543': 4.45,
 '106': 4.44,
 '319': 4.43,
 '601': 4.43,
 '413': 4.41,
 '475': 4.41,
 '188': 4.4,
 '12': 4.39,
 '276': 4.39,
 '154': 4.38,
 '533': 4.38,
 '581': 4.38,
 '1': 4.37,
 '69': 4.37,
 '586': 4.37,
 '59': 4.36,
 '128': 4.36,
 '544': 4.36,
 '253': 4.35,
 '459': 4.35,
 '553': 4.34,
 '585': 4.34,
 '519': 4.33,
 '70': 4.32,
 '336': 4.32,
 '435': 4.32,
 '300': 4.3,
 '93': 4.29,
 '532': 4.28,
 '74': 4.27,
 '49': 4.26,
 '80': 4.26,
 '224': 4.26,
 '291': 4.26,
 '164': 4.25,
 '169': 4.25,
 '246': 4.25,
 '398': 4.25,
 '162': 4.24,
 '209': 4.24,
 '362': 4.24,
 '494': 4.23,
 '250': 4.22,
 '17': 4.21,
 '337': 4.21,
 '340': 4.21,
 '573': 4.21,
 '79': 4.2,
 '227': 4.2,
 '252': 4.2,
 '258': 4.2,
 '460': 4.2,
 '595': 4.2,
 '97': 4.19,
 '364': 4.19,
 '526

In [14]:
%timeit Ratings().Users().top_by_ratings(metric='median')
top_users_by_median = Ratings().Users().top_by_ratings(metric='median')
top_users_by_median

97.8 ms ± 5.21 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


{'1': 5.0,
 '25': 5.0,
 '30': 5.0,
 '43': 5.0,
 '52': 5.0,
 '53': 5.0,
 '59': 5.0,
 '69': 5.0,
 '77': 5.0,
 '106': 5.0,
 '154': 5.0,
 '164': 5.0,
 '171': 5.0,
 '188': 5.0,
 '224': 5.0,
 '236': 5.0,
 '251': 5.0,
 '258': 5.0,
 '267': 5.0,
 '276': 5.0,
 '348': 5.0,
 '393': 5.0,
 '413': 5.0,
 '417': 5.0,
 '441': 5.0,
 '452': 5.0,
 '494': 5.0,
 '515': 5.0,
 '523': 5.0,
 '533': 5.0,
 '543': 5.0,
 '544': 5.0,
 '598': 5.0,
 '12': 4.75,
 '49': 4.5,
 '70': 4.5,
 '72': 4.5,
 '88': 4.5,
 '122': 4.5,
 '168': 4.5,
 '169': 4.5,
 '178': 4.5,
 '209': 4.5,
 '210': 4.5,
 '221': 4.5,
 '246': 4.5,
 '250': 4.5,
 '253': 4.5,
 '295': 4.5,
 '296': 4.5,
 '319': 4.5,
 '336': 4.5,
 '371': 4.5,
 '398': 4.5,
 '399': 4.5,
 '400': 4.5,
 '421': 4.5,
 '435': 4.5,
 '459': 4.5,
 '460': 4.5,
 '475': 4.5,
 '495': 4.5,
 '505': 4.5,
 '537': 4.5,
 '538': 4.5,
 '548': 4.5,
 '553': 4.5,
 '573': 4.5,
 '578': 4.5,
 '581': 4.5,
 '585': 4.5,
 '586': 4.5,
 '594': 4.5,
 '601': 4.5,
 '138': 4.25,
 '327': 4.25,
 '2': 4.0,
 '4': 4.0,
 '

### Analysis of user-generated ratings employing measures of central tendency (mean and median) revealed a systematic propensity for users to assign scores marginally above the midpoint of the available rating scale. This observed positive skew in rating distributions suggests a consistent behavioral tendency toward lenient evaluation patterns among the user population.

### Then let's examine users' rating behaviour and find the most controversial users.

In [15]:
%timeit Ratings().Users().top_controversial(10)
contr_users= Ratings().Users().top_controversial(10)
contr_users

113 ms ± 7.89 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


{'3': 4.37,
 '55': 3.22,
 '461': 3.22,
 '259': 3.05,
 '329': 3.05,
 '175': 2.87,
 '502': 2.84,
 '598': 2.84,
 '393': 2.63,
 '138': 2.56}

### The analysis reveals maximum rating variance (4.37) for user ID 3, suggesting this individual exhibits polarized rating behavior. This statistical pattern implies the user predominantly evaluates cinematic content that elicits extreme emotional responses, either strongly positive or negative, rather than producing neutral assessments.

## &emsp;&emsp;&emsp;&emsp;  Tag Ecosystem Exploration

### Now let's analyze users' activity in terms of creating tags for movies.

In [16]:
%timeit Tags().most_words(10)
most_words = Tags().most_words(10)
most_words

4.73 ms ± 272 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'something for everyone in this one... saw it without and plan on seeing it with kids!': 16,
 'the catholic church is the most corrupt organization in history': 10,
 'villain nonexistent or not needed for good story': 8,
 'it was melodramatic and kind of dumb': 7,
 '06 oscar nominated best movie - animation': 7,
 'stop using useless characters for filler': 6,
 'oscar (best effects - visual effects)': 6,
 'oscar (best music - original score)': 6,
 'give me back my son!': 5,
 'based on a true story': 5}

In [17]:
%timeit Tags().longest(10)
most_longest = Tags().longest(10)
most_longest

4.42 ms ± 210 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


['something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'villain nonexistent or not needed for good story',
 'r:disturbing violent content including rape',
 '06 oscar nominated best movie - animation',
 'stop using useless characters for filler',
 'academy award (best supporting actress)',
 'oscar (best effects - visual effects)',
 'it was melodramatic and kind of dumb',
 'r:sustained strong stylized violence']

### The longest tags (both in terms of words' number and symbols' number) capture detailed sentiments or emotional reactions of users to specific movie.

In [18]:
%timeit Tags().most_words_and_longest(10)
most_words_and_longest = Tags().most_words_and_longest(10)
most_words_and_longest

5.07 ms ± 110 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


['oscar (best effects - visual effects)',
 'the catholic church is the most corrupt organization in history',
 'villain nonexistent or not needed for good story',
 '06 oscar nominated best movie - animation',
 'stop using useless characters for filler',
 'something for everyone in this one... saw it without and plan on seeing it with kids!',
 'it was melodramatic and kind of dumb']

### These tags highlight expressive users describing cinematography.

In [19]:
%timeit Tags().most_popular(100)
most_popular = Tags().most_popular(100)
most_popular

4.45 ms ± 240 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'in netflix queue': 131,
 'atmospheric': 41,
 'funny': 24,
 'superhero': 24,
 'surreal': 24,
 'thought-provoking': 24,
 'sci-fi': 23,
 'disney': 23,
 'quirky': 22,
 'religion': 22,
 'dark comedy': 21,
 'suspense': 21,
 'psychology': 21,
 'twist ending': 20,
 'visually appealing': 20,
 'comedy': 19,
 'politics': 19,
 'crime': 19,
 'music': 17,
 'dark': 16,
 'action': 16,
 'time travel': 16,
 'high school': 16,
 'mental illness': 16,
 'mindfuck': 15,
 'aliens': 15,
 'black comedy': 14,
 'space': 14,
 'dreamlike': 14,
 'mafia': 13,
 'holocaust': 13,
 'disturbing': 13,
 'shakespeare': 13,
 'emotional': 13,
 'stephen king': 13,
 'journalism': 13,
 'heist': 13,
 'classic': 12,
 'anime': 12,
 'satire': 12,
 'court': 12,
 'christmas': 12,
 'animation': 11,
 'imdb top 250': 11,
 'comic book': 11,
 'psychological': 11,
 'ghosts': 11,
 'bittersweet': 11,
 'adolescence': 11,
 'adultery': 11,
 'boxing': 11,
 'drugs': 10,
 'leonardo dicaprio': 10,
 'robots': 10,
 'cinematography': 10,
 'india': 10,

### This list of user-generated tags reveals a fascinating blend of genres, themes, and viewing preferences, reflecting a diverse audience with eclectic tastes. 

### The top tag, "in netflix queue" (131), overshadows others, suggesting users prioritize tracking films they plan to watch rather than categorizing by genre or mood.

In [20]:
%timeit Tags().tags_with('Netflix')
tags_with = Tags().tags_with('Netflix')
tags_with

3.67 ms ± 148 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


['in netflix queue', 'no dvd at netflix', 'not available from netflix']

### The diversity of tags with "Netflix" additionally emphasizes the popularity of this streaming platform for users.

## &emsp;&emsp;&emsp;&emsp; Financial & Production Context (via IMDB)

### Analyze prolific and famous directors.

In [21]:
%timeit Links().top_directors(10)
top_directors = Links().top_directors(10)
top_directors

3.49 s ± 226 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


{'Alfred Hitchcock': 11,
 'Woody Allen': 8,
 'Stanley Kubrick': 6,
 'Frank Capra': 6,
 'John Lasseter': 5,
 'Rob Reiner': 5,
 'Martin Scorsese': 5,
 'Tony Scott': 5,
 'John Carpenter': 5,
 'James Cameron': 5}

### This list of top directors reflects a blend of legendary auteurs, genre pioneers, and mainstream hitmakers, spanning decades of cinema history.
### Alfred Hitchcock is the undisputed leader (see below).
### Overall, this list leans heavily on canonical, male-dominated directors from the 20th century, emphasizing timelessness over contemporary trends. 

In [22]:
%timeit Links().all_movies_by_director(top_directors, 'Alfred Hitchcock')
all_movies_by_dir = Links().all_movies_by_director(top_directors, 'Alfred Hitchcock')
all_movies_by_dir

18.6 ms ± 584 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


['Vertigo',
 'Rear Window',
 'North by Northwest',
 'Rebecca',
 'Foreign Correspondent',
 'Notorious',
 'Spellbound',
 'To Catch a Thief',
 'The 39 Steps',
 'Dial M for Murder',
 'Psycho']

### Alfred Hitchcock (11): The undisputed leader, Hitchcock’s prominence underscores his timeless influence as a director.
### Films like Psycho and Vertigo remain cornerstones of thriller/suspense genres, and his work is frequently studied for its psychological depth and technical innovation.

### Next, find High-budget films. (without taking into account the inflation rate)

In [23]:
%timeit Links().most_expensive(10)
most_expensive = Links().most_expensive(10)
most_expensive

7.41 s ± 225 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


{'Waterworld': 175000000,
 'True Lies': 115000000,
 'Terminator 2: Judgment Day': 102000000,
 'Batman Forever': 100000000,
 'The Hunchback of Notre Dame': 100000000,
 'Eraser': 100000000,
 'Cutthroat Island': 98000000,
 'Twister': 92000000,
 'Die Hard with a Vengeance': 90000000,
 'Judge Dredd': 90000000}

### This list of high-budget films offers a snapshot of mid-1990s Hollywood blockbuster culture, marked by ambitious spending, genre trends, and mixed outcomes. 
### Inflation-adjusted costs: while budgets are already massive (e.g., Waterworld at 175M in 1995≈360M today), these films pushed financial boundaries for their time, driven by rising special effects and star salaries.

###  Analyze films from the dataset, which have the most box office success, the longest runtime and the highest cost per minute calculated.

In [24]:
%timeit Links().most_profitable(10)
most_profitable = Links().most_profitable(10)
most_profitable

9.11 s ± 834 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


{'Jimmy Hollywood': 857100000.0,
 'The Doors': 782465500.0,
 'Kiss of Death': 764398007.0,
 'Trainspotting': 742400891.0,
 'The Jungle Book': 718455561.0,
 'Cobb': 622387716.0,
 'Lawrence of Arabia': 540350000.0,
 'Tie Me Up! Tie Me Down!': 520400000.0,
 'The Brady Bunch Movie': 483000000.0,
 'Home Alone': 476050219.0}

### "Jimmy Hollywood" (1994) is a notorious box-office bomb.

In [25]:
%timeit Links().longest(10)
top_longest = Links().longest(10)
top_longest

5.38 s ± 405 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


{'Emma': '240 min',
 'Gone with the Wind': '238 min',
 'Once Upon a Time in America': '229 min',
 'Lawrence of Arabia': '227 min',
 'Ben-Hur': '212 min',
 'The Godfather Part II': '202 min',
 'Giant': '201 min',
 "Schindler's List": '195 min',
 'The Right Stuff': '193 min',
 'Nixon': '192 min'}

### This list of films paired with their runtimes highlights a collection of epic-scale cinema, characterized by ambitious storytelling, historical depth, and often monumental production values. 
### Common themes & genres are Historical Epics, Biographical Dramas and Crime Sagas.

In [26]:
%timeit Links().top_cost_per_minute(5)
top_cost_per_minute = Links().top_cost_per_minute(5)
top_cost_per_minute

7.99 s ± 782 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


{'Waterworld': 1296296.3,
 'The Hunchback of Notre Dame': 1098901.1,
 'Judge Dredd': 937500.0,
 'Space Jam': 909090.91,
 'Eraser': 869565.22}

### This list of "top cost per minute" films—now featuring Waterworld, The Hunchback of Notre Dame, Judge Dredd, Space Jam, and Eraser—paints a clearer picture of mid-1990s Hollywood excess, where studios gambled heavily on spectacle, star power, and untested concepts. 
### These films exemplify 1990s Hollywood’s "go big or go home" mentality—where budgets ballooned around stars, effects, and IP, often without disciplined oversight. 
### If these films were made today, their budgets would approach $200–300M+ (adjusted for inflation), making their struggles even more glaring.